Avoid unnecessary monadic lifting of partial applications

MajaRet opened the pull request:
Previously, the compiler performed eta abstractions any time it encountered a partial
application on a function declaration's right-hand side. These abstractions were done
by transforming the partial application into a lambda abstraction that includes the
missing arguments.

When this happens on the top level (i.e. a function is defined as a partial application
of another function or constructor), this leads to an additional lifting into the Free
monad.
This can be avoided by applying the right-hand side to the missing arguments and
adding them to the left-hand side of the function rule explicitly.

The change is implemented in the compiler's `EtaConversionPass`.

Closes #3

Commit:
Avoid unnecessary monadic lifting when functions are defined as partial applications #3

Previously, the compiler performed eta abstractions any time it encountered a partial
application on a function declaration's right-hand side. These abstractions were done
by transforming the partial application into a lambda abstraction that includes the
missing arguments.

When this happens on the top-level (i.e. a function is defined as a partial application
of another function or constructor), this leads to an additional lifting into the Free
monad.
This can be avoided by applying the right-hand side to the missing arguments and
adding them to the left-hand side of the function rule explicitly.

This commit implements this optimization.

Commit:
Remove import of Pretty

Commit:
Format code with Brittany

Commit:
Merge branch 'master' of https://github.com/FreeProving/free-compiler into issue-3

MajaRet commented:
I'm not sure why the Brittany check fails. Running the check-formatting.sh script changed 22 files for some reason (hence the 22 changed files), but the check still fails.

Commit:
Revert "Format code with Brittany"

This reverts commit a6e56f2069d26877cf4e71930ca6bd7dc7c40fe2.

Commit:
Reformat EtaConversionPass.hs with Brittany #3

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -31,17 +44,27 @@
+--
+--   where @g@ is the name of an @n@-ary constructor or function declaration and 
+--   @m < n@. This declaration is then replaced by 
+--   > f e₁ … eₖ x₍ₘ₊₁₎ … xₙ = @α₁ … @αₚ g e₁ … eₘ x₍ₘ₊₁₎ … xₙ
```
Just a small typo.

```suggestion
--   > f e₁ … eₖ x₍ₘ₊₁₎ … xₙ = g @α₁ … @αₚ e₁ … eₘ x₍ₘ₊₁₎ … xₙ
```

Very good job on the documentation btw!

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,27 +104,67 @@ etaConversionPass ast = do
+  let rhs = IR.funcDeclRhs funcDecl
+  xs   <- generateFullArgumentList rhs
+  rhs' <- etaConvertTopLevelRhs xs rhs
+  -- compute the function's new (uncurried) type
```
All comments should be proper sentences.

```suggestion
  -- Compute the function's new (uncurried) type.
```

Commit:
Modify EtaConversionPass.hs to allow testing new behavior #3

Changes:
- Make `etaConvertTopLevelRhs` the testing interface instead of
  `etaConvertExpression`, because the new behavior cannot be tested
  with the latter.
- Pass a list of `VarPat`s to `etaConvertTopLevelRhs` instead of passing
  a list of strings (identifiers) and then translating them to `VarPat`s.
- Get the return type information from the environment instead of using
  IR.funcDeclReturnType. Before, the unit tests failed because the
  IR.funcDeclReturnType field is always `Nothing` unless the other passes
  (namely type inference) are executed first, which is not done during
  testing.

Commit:
Modify testing interface for the EtaConversionPass #3

Since the EtaConversionPass may now modify the left-hand side
of a rule as well as the right-hand side, the whole function
declaration should now be tested instead of only the
right-hand side.
Therefore, the testing interface has been changed and the
unit tests in EtaConversionPassTests modified to use the new
interface.

I have also added an example test case for reference.

The old tests don't work anymore, but I have left them in the
file as comments for reference. They should be removed later
after the new tests have been added.

Commit:
Fix errors in test case for EtaConversionPass #3

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,27 +104,69 @@ etaConversionPass ast = do
+  expr' <- etaConvertSubExprs expr
+  if null argPats
+    then return expr'
+    else return (IR.app NoSrcSpan expr' argExprs)
```
You don't need to test whether there are any arguments. `IR.app` returns the expression if the argument list is empty.

```suggestion
  return (IR.app NoSrcSpan expr' argExprs)
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,27 +104,69 @@ etaConversionPass ast = do
+  if null argPats
+    then return expr'
+    else return (IR.app NoSrcSpan expr' argExprs)
+  where argExprs = map IR.varPatToExpr argPats
```
Since you have a `do` block above, I'd use a `let` statement in this case.

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,27 +104,69 @@ etaConversionPass ast = do
-  return funcDecl { IR.funcDeclRhs = rhs' }
+  let rhs = IR.funcDeclRhs funcDecl
+  xs          <- generateFullArgumentList rhs
+  returnType' <- inEnv
+    $ lookupReturnType IR.ValueScope (IR.funcDeclQName funcDecl)
```
I'd recommend using the return type annotation from `funcDecl` directly.

MajaRet replied:
I did that at first, but then the unit tests didn't work because apparently, `IR.funcDeclReturnType` contains `Nothing` when you define a test function.

just95 replied:
Okay, that makes sense. However, in the test cases you could just annotate the return type and add to the preconditions of the pass that the function declarations must be type annotated.

```haskell
  "f :: Foo -> Foo = g Foo" `shouldEtaConvert` "f (y :: Foo) :: Foo = g Foo y"
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,27 +104,69 @@ etaConversionPass ast = do
+                           }
+        )
+        args
+        xs
```
This should be equivalent to the following.

```suggestion
  let newArgs = zipWith (IR.VarPat NoSrcSpan) xs (map Just args)
```

A better name for `args` could be `newArgTypes` and for `xs` a name like `newArgIdents` would be more descriptive.

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -169,3 +238,26 @@ arityOf (IR.Undefined _ _      ) = return 0
+-- Helper functions                                                                     --
+-------------------------------------------------------------------------------
+
+-- | 'splitReturnType n' uncurries the given function's type n times.
+--   It is assumed that the given type is of the form
+--   @τ₁ -> (… -> (τₙ -> τ))@, which is true for the return type of 
+--   a partially defined function with @n@ missing arguments.
+--   The function returns a tuple consisting of a list of the argument
+--   types @[τ₁, …, τₙ]@ and the return type @τ@. 
+splitReturnType :: Int -> IR.Type -> ([IR.Type], IR.Type)
+splitReturnType 0 t = ([], t)
+splitReturnType n t =
+  let (arg , r    ) = IR.splitFuncType t 1
+      (args, rType) = splitReturnType (n - 1) r
+  in  (arg ++ args, rType)
```
I don't get how this is different from `IR.splitFuncType` (except that the arguments are flipped).

Also, the assumption that the type has the form `τ₁ -> (… -> (τₙ -> τ))` is not necessarily true in the presence of type synonyms. The [`TypeSignaturePass`](https://github.com/FreeProving/free-compiler/blob/4627f930212624096a61ad051b557c0690f4d838/src/lib/FreeC/Pass/TypeSignaturePass.hs#L187) deals with this issue by expanding type synonyms on demand. I think that the most reliable solution would be to move the `etaConversionPass` above the `typeSignaturePass` in the compiler [`pipeline`](https://github.com/FreeProving/free-compiler/blob/4627f930212624096a61ad051b557c0690f4d838/src/lib/FreeC/Pipeline.hs#L25). This would have the advantage, that you wouldn't have to annotate any types or modify environment entries. However, this change would be more involved and is not really in the scope of a first issue. Thus, I'd suggest that we just recycle the definition of `FreeC.Pass.TypeSignaturePass.splitFuncType` for now.

MajaRet replied:
It isn't; I simply misunderstood what `IR.splitFuncType` actually does. I've changed it so that `etaConvertFuncDecl` simply uses `IR.splitFuncType` now.

just95 replied:
Ah, okay. If you have a suggestion for how to improve the documentation of `IR.splitFuncType` (or any other function), don't hesitate to rephrase it : )

just95 commented on src/test/FreeC/Pass/EtaConversionPassTests.hs:
```diff
@@ -32,6 +33,20 @@ shouldEtaConvert inputStr expectedOutputStr = do
+        _ <- defineTestFunc "g" 2 "Foo -> Foo -> Foo"
+        "f = g Foo" `shouldEtaConvert` "f (y :: Foo) :: Foo = g Foo y"
+
+-- These are the old tests, but since the testing interface has changed
+-- and we now test whole function declarations instead of right-hand sides,
+-- these tests don't work anymore.
+-- I'll leave them here for reference for now, but they should be removed later.
+{-
```
All of the old test cases should still work. I'd suggest to export `etaConvertExpr` again and move the new and existing test cases to their own [`context`](https://hackage.haskell.org/package/hspec-2.7.1/docs/Test-Hspec.html#v:context)es.

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,27 +104,69 @@ etaConversionPass ast = do
+  rhs'       <- etaConvertTopLevelRhs newArgs rhs
+  -- update the environment with the new type and arguments
+  Just entry <- inEnv $ lookupEntry IR.ValueScope (IR.funcDeclQName funcDecl)
+  modifyEnv $ addEntry entry { entryArity      = length vars'
```
Good catch that the environment entries need to be updated : )

However, when the arity of a function in the environment changes, additional η-conversions become necessary. Therefore, η-conversion has to be performed recursively.

MajaRet committed:
Fix typo in EtaConversionPass

Co-authored-by: Justin Andresen <jan.dresen.95@gmail.com>

Commit:
Apply suggestions #3

- Rename args to newArgTypes and xs to newArgIdents.
- Simplify some expressions.
- Remove splitReturnType and use IR.splitFuncType instead.
- Get the return type through IR.funcDeclReturnType, not from the environment.
- Re-add etaConvertExpr to the testing interface.
- Re-add the old test cases in their own context.
- Add the precondition that all function declarations must be type annotated.

Commit:
Apply HLint suggestion #3

just95 commented on src/test/FreeC/Pass/EtaConversionPassTests.hs:
```diff
@@ -18,47 +18,70 @@ import           FreeC.Test.Expectations
+      $                  shouldSucceedWith
+      $                  do
+                           "\\(f :: a -> b -> c) x -> f x"
       `shouldEtaConvert` "\\(f :: a -> b -> c) x -> f x"
```
In this case it probably better to write the expectation in prefix notation such that Brittany does not mess up the formatting.

ichistmeinname reviewed (approved):
I really like the effort you put into the documentation, great work!

Commit:
Reuse TypeSignaturePass's definition of splitFuncType to handle type synonyms #3

Commit:
Also perform modified eta conversion in top-level case and if expressions #3

Commit:
Add test for eta conversion applied to mutually recursive functions

Commit:
Add draft for similar instance for modules

ichistmeinname commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -237,8 +317,15 @@ arityOf (IR.Lambda _ _ _ _     ) = return 0
+minM m1 m2 = do
+  i1 <- m1
+  i2 <- m2
+  return (if i1 <= i2 then i1 else i2)
```
It's a pity that there's no `minimumM` or `minimumA` function in some Haskell library. Nevertheless, as your version of `minM` is basically a lifted version of `min`, we can rewrite it as one fold that yields `return 0`. I'd suggest to then outsource this function into a `Util` module that should exist somewhere already (ping @just95). For that sake, I generalised the function from `Converter Int` to `m a`, where  `a` is `Ord` (for the usage of `min`) and `Bounded` (in order to yield the smallest value otherwise).

```suggestion
minimumM :: (Ord a, Bounded a) => [m a] -> m a
minimumM = foldr (liftM2 min) (return minBound)
```

ichistmeinname commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -209,6 +276,19 @@ etaConvertSubExprs' expr = do
+--   All non-zero arities on the right-hand side of a function declaration 
+--   have the same arity.
+findMinMissingArguments :: IR.Expr -> Converter Int
+findMinMissingArguments (IR.If _ _ e1 e2 _) =
+  minM (findMinMissingArguments e1) (findMinMissingArguments e2)
```
We can get rid of calling `findMinMissingArguments` twice!

```suggestion
findMinMissingArguments :: IR.Expr -> Converter Int
findMinMissingArguments (IR.If _ _ e1 e2 _) =
  minimumM (map findMinMissingArguments [e1, e2])
```

ichistmeinname commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -106,57 +132,98 @@ etaConversionPass ast = do
+  etaConvertExpr $ IR.app NoSrcSpan expr argExprs
+
+-- | Calls @etaConvertTopLevel@ on all alternatives in an if or case expression. 
+etaConvertTopLevel' :: [IR.VarPat] -> IR.Expr -> Converter IR.Expr
```
Can we think of a better name here?

ichistmeinname commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -106,57 +132,98 @@ etaConversionPass ast = do
+  -- sides are missing at least one argument.
+  if newArgNumber == 0
+    then do
+      rhs' <- etaConvertExpr (IR.funcDeclRhs funcDecl)
```
```suggestion
      rhs' <- etaConvertExpr rhs
```

ichistmeinname commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -106,57 +132,98 @@ etaConversionPass ast = do
-  let (newArgTypes, returnType) = IR.splitFuncType
-        (fromJust $ IR.funcDeclReturnType funcDecl)
-        (length newArgIdents)
+  newArgNumber <- findMinMissingArguments rhs
```
I'm currently regretting reading the file bottom to top instead of the other way around : ) I now have the impression that `findMinMissingArguments` and `arityOf` do not need to be monadic (i.e., of type `Converter <something>`) -- there's not context needed to compute the needed value. Can you confirm or refute that impression?

MajaRet replied:
`arityOf` needs to look up the arity of a function in the current environment, so it does need to produce a value inside the `Converter` monad. 
`inEnv` has the type `MonadConverter m => (Environment -> a) -> m a`. 
Since `findMinMissingArguments` uses `arityOf`, it also needs to produce a monadic value.

ichistmeinname replied:
Ah, I see. I missed that from the latest diff I was looking at!

MajaRet committed:
Apply some suggestions from code review #3

Co-authored-by: Sandra Dylus <sad@informatik.uni-kiel.de>

Commit:
Fix compiler errors #3

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -142,10 +272,24 @@ etaConvertSubExprs' expr = do
 -- Arity                                                                     --
 -------------------------------------------------------------------------------
 
+-- | Find the minimum number of missing arguments (the arity of the expressions) among the alternatives
+--   for the right-hand side of a function declaration. 
+--   All non-zero arities on the right-hand side of a function declaration 
+--   have the same arity.
+findMinMissingArguments :: IR.Expr -> Converter Int
+findMinMissingArguments (IR.If _ _ e1 e2 _) =
+  minimumM (map findMinMissingArguments [e1, e2])
+findMinMissingArguments (IR.Case _ _ alts _) =
+  minimumM $ map (findMinMissingArguments . IR.altRhs) alts
```
We can avoid to add `minimumM` and a monad utility module by writing the following.

```haskell
minimum <$> mapM (findMinMissingArguments . IR.altRhs) alts
```

MajaRet replied:
Good point!

ichistmeinname replied:
I had a version using `liftM` as well but refrained from this suggestion afterwards ; ) So, yes, please go ahead with this suggestion!

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,20 +98,30 @@
 import qualified FreeC.IR.Syntax               as IR
 import           FreeC.Monad.Converter
 import           FreeC.Pass
+-- temporary import; ideally, this pass should be moved before the TypeSignaturePass.
+import           FreeC.Pass.TypeSignaturePass   ( splitFuncType )
```
Could you make this a `TODO` comment such that we don't forget that we have to address this in the future?

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,18 +135,91 @@ etaConversionPass ast = do
+  let rhs = IR.funcDeclRhs funcDecl
+  newArgNumber <- findMinMissingArguments rhs
+  -- Only perform top-level eta conversion when all alternatives for right-hand sides are missing at least one argument.
+  if newArgNumber == 0
+    then do
+      rhs' <- etaConvertExpr rhs
+      return funcDecl { IR.funcDeclRhs = rhs' }
```
Shouldn't the `else` branch cover this case already?

MajaRet replied:
Yes, you're right. I thought this way, we'd avoid unnecessary function evaluations and the 'updating' of the environment, but thinking about it again, it just causes different unnecessary function evaluations (like computing the arity of the right-hand side again).

Commit:
Apply suggestions from code review #3

jo-hanna1997 reviewed (commented):
Your implementation looks good and I like the extension for if-expressions on top-level.
An interesting case occurs when in both if-branches functions with different arity occur.
For example:
```
--assume g as arity 2
g :: Foo -> Foo -> Foo 
--assume h as arity 1
h:: Foo -> (Foo -> Foo)

f :: Bool -> Foo -> Foo
f b = if b then g Foo else h Foo
```
Currently this will be transformed into:

`f b = if b then (\x -> g Foo x) else h Foo`

But we could also over apply `h` which is no harm to obtain

`f b x = if b the g Foo x else h Foo x` 

Furthermore I do think a recursive call is neccessary if we consider mutually recursive functions.
Let's say we have `g` as above and 
```
k :: Foo -> Foo
k = f

f :: Foo -> Foo
f = g (k Foo)
```
Then a first run would update `f` to `f x = g (k Foo) x`.
Now `f` has arity 1 but in `k` `f` is applied to no arguments. 
With repeated runs this should be resolved and yield
`k = (\x -> f x)`

ichistmeinname commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,18 +133,83 @@ etaConversionPass ast = do
+etaConvertAlternatives argPats expr = do
+  -- The first child term of an if or case expression is the condition/the scrutinee
+  -- and should remain unchanged.
+  let (e : subterms) = childTerms expr
```
Are we sure that this binding can never fail, that is, `childTerms expr` never yields an empty list?

MajaRet replied:
It should never yield an empty list. `etaConvertAlternatives` is only used when `expr` is an if or case expression, and `childTerms` is defined as follows for those kinds of expressions (in `FreeC.IR.Subterm`):
```
  childTerms (IR.If _ e1 e2 e3 _        ) = [e1, e2, e3]
  childTerms (IR.Case   _ expr alts _   ) = expr : map IR.altRhs alts
```
So the list always has at least one element.

just95 replied:
I agree that it is safe to assume that the list cannot be empty. The documentation of `childTerms` states that the list contains all direct child nodes. Since `if` and `case` expressions both have at least one direct child, the list is guaranteed  to be non-empty by the specification in this situation. However, `childTerms`  does not really provide any guarantee that the nodes are sorted from left to right. That's just how it happens to be implemented right now. Thus, I would recommend to perform the recursion in `etaConvertTopLevel` explicitly.

MajaRet replied:
Okay, that makes sense. I'll change it then.

ichistmeinname commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -78,18 +133,83 @@ etaConversionPass ast = do
+  -- and should remain unchanged.
+  let (e : subterms) = childTerms expr
+  subterms' <- mapM (etaConvertTopLevel argPats) subterms
+  let Just expr' = replaceChildTerms expr (e : subterms')
```
Maybe we should discuss, if we want to just `fromJust` directly or use the let-binding like here. Any arguments on your or @just95's side?

just95 replied:
In this case, I'd keep it as is for now. The issue will be addressed in #52 anyway.

Commit:
WIP

Commit:
Fix mistake where the EtaConversionPass was not executed #3

I stopped the EtaConversionPass from being executed for testing
purposes, then forgot to undo that.
I undid it now, so the pass is executed again.

Commit:
Perform eta conversion recursively if the type of a function is changed #3

Commit:
Merge branch 'issue-3' of https://github.com/FreeProving/free-compiler into issue-3

Commit:
Format code with Brittany #3

Commit:
Write expectation in prefix notation if arguments are too long #3

Commit:
Merge branch 'master' of https://github.com/FreeProving/free-compiler into issue-3

Commit:
Adapt to the new environment entries without Maybe #3

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,45 +97,133 @@
+--   previously converted function declarations. 
+--   This ensures that all functions, including mutually-recursive 
+--   functions, are fully applied correctly. 
+etaConvertFuncDecls :: [IR.FuncDecl] -> [IR.FuncDecl] -> Converter [IR.FuncDecl]
```
It would be nice if you could add an explanation of the function arguments to the Haddock documentation.

See [Function Arguments](https://haskell-haddock.readthedocs.io/en/latest/markup.html#function-arguments).

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,45 +97,133 @@
 
--- | Applies 'etaConvertExpr' to the right-hand side of the given function
---   declaration.
+-- | Make sure all occurring functions are fully applied by calling 'etaConvertFuncDecl' on each function declaration. 
```
Documentation of function declarations should start with a third-person verb.

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,45 +97,133 @@
+    then etaConvertFuncDecls (newFuncDecls ++ (newFuncDecl : fds)) []
+    else etaConvertFuncDecls fds (newFuncDecls ++ [newFuncDecl])
+
+-- | Depending on the presence or absence of missing top-level arguments, applies 'modifyTopLevel' or 'etaConvertExpr' to the right-hand side of the given function declaration to ensure all functions 
```
This line looks way to long. Make sure no comments exceed the maximum line length of 80 characters.

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,45 +97,133 @@
+  subterms' <- mapM (etaConvertTopLevel argPats) subterms
+  let Just expr' = replaceChildTerms expr (e : subterms')
+  return expr'
+
+
```
There should be at most one blank line between two code blocks. There are multiple redundant blank lines in this file and the tests that can be removed.

just95 reviewed (changes requested):
Looks good to me. There are just some formatting and documentation related issues that I would like you to address before we merge this.

jo-hanna1997 commented:
It would be nice to know how you decided to proceed, converning the comment I wrote, so that I can adjust the tests.

MajaRet commented:
> It would be nice to know how you decided to proceed, converning the comment I wrote, so that I can adjust the tests.

Yes, of course, sorry. I added the recursion (by introducing a new function `etaConvertFuncDecls`). I also made the other change you suggested, but I did it in another branch called 'issue-3-overapplied' because I'm not completely sure that the version with over-applications is always equivalent in the presence of effects. 
All that changed there is that the two occurrences of `minimum` were changed to `maximum`.

Commit:
Order imports and correct line endings

Commit:
Merge branch 'master' into issue-3-test

Commit:
Improve documentation #3

Commit:
Merge branch 'master' of https://github.com/FreeProving/free-compiler into issue-3

Commit:
Add the strictness flag argument to application of IR.VarPat #3

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -22,26 +22,72 @@
+--  >           c₂ -> h 
 --
---   where @x₍ₘ₊₁₎ … xₙ@ are @n-m@ fresh variables.
+--  where c₁ and c₂ are the constructors of τ₁, 
```
```suggestion
--  where @c₁@ and @c₂@ are the constructors of @τ₁@, 
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,45 +97,142 @@
 import           FreeC.Pass
+-- temporary import; TODO: move this pass before the TypeSignaturePass.
+import           FreeC.Pass.TypeSignaturePass   ( splitFuncType )
+
```
```suggestion
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -22,26 +22,72 @@
+--   where x₍ₘ₊₁₎ … xₙ are @n-m@ fresh variables. 
+--
+--   If a function rule has several alternatives for the right-hand side through 
+--   -- possibly nested -- if or case expressions, the number of added arguments 
```
Are the `--` intentional (en- or em-dashes) or has something gone wrong during formatting? I'd suggest to use Unicode (i.e., `–` or `—`) if you want to insert a dash.

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -22,26 +22,72 @@
+--   Both types of η-conversion may also be applied to the same expression.
+--   For example, the function 
+--   
+--  > f :: τ₁ -> τ₂ -> τ₃ > τ 
+--  > f e₁ = case e₁ of
+--  >           c₁ -> g 
+--  >           c₂ -> h 
```
Maybe IR notation should be used here.

```suggestion
--  > f :: τ₁ -> τ₂ -> τ₃ > τ 
--  > f e₁ = case e₁ of {
--  >     c₁ -> g;
--  >     c₂ -> h 
--  >   }
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -146,6 +289,17 @@ etaConvertSubExprs' expr = do
 -- Arity                                                                     --
 -------------------------------------------------------------------------------
 
+-- | Find the minimum number of missing arguments among the alternatives
```
```suggestion
-- | Finds the minimum number of missing arguments among the alternatives
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,45 +97,142 @@
+
+-- | Applies appropriate η-conversions to a function declaration.
+--
+-- Depending on the presence or absence of missing top-level arguments,
+-- the function uses 'etaConvertTopLevel' or 'etaConvertExpr' to ensure 
+-- all functions and constructors on the right-hand side are fully applied. 
+-- The missing top-level arguments are also added to the left-hand
+-- side of the declaration and the function's type and the environment
+-- are updated accordingly. 
```
This paragraph should be aligned with the subject line. There is another line in 207 that needs to be aligned.
After aligning the comments, make sure that the 80 character limit is still met. Some lines are still slightly too long right now.

```suggestion
--   Depending on the presence or absence of missing top-level arguments,
--   the function uses 'etaConvertTopLevel' or 'etaConvertExpr' to ensure 
--   all functions and constructors on the right-hand side are fully applied. 
--   The missing top-level arguments are also added to the left-hand
--   side of the declaration and the function's type and the environment
--   are updated accordingly. 
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -51,45 +97,142 @@
+-- are fully applied.
+etaConvertTopLevel argPats expr = localEnv $ do
+  let argExprs = map IR.varPatToExpr argPats
+  -- apply expression to missing arguments and perform eta-conversion on the 
+  -- resulting expression 
```
```suggestion
  -- Apply expression to missing arguments and perform eta-conversion on the 
  -- resulting expression.
```

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -22,26 +22,72 @@
+--
+--   where @g@ is the name of an @n@-ary constructor or function declaration and 
+--   @m < n@. This declaration is then replaced by 
+--   > f e₁ … eₖ x₍ₘ₊₁₎ … xₙ = g @α₁ … @αₚ e₁ … eₘ x₍ₘ₊₁₎ … xₙ
```
Some markup languages have problems with consecutive block-lvel elements. I don't know how Haddock handles this case but I'd insert a blank line between block-level elements just to be safe.

```suggestion
--
--   > f e₁ … eₖ x₍ₘ₊₁₎ … xₙ = g @α₁ … @αₚ e₁ … eₘ x₍ₘ₊₁₎ … xₙ
```

jo-hanna1997 committed:
Change tests to expect minimum argument number paradigm for if-expressions

jo-hanna1997 committed:
Merge branch 'master' into issue-3-test

Commit:
Fix merge conflict

Commit:
Remove maybe syntax and correct test

Commit:
Apply hlint suggestions and change line endings

Commit:
Apply formatting suggestions #3

jo-hanna1997 committed:
Apply brittany formatting

MajaRet committed:
Merge pull request #65 from FreeProving/issue-3-test

Issue 3 test

Commit:
Merge branch 'issue-3' of https://github.com/FreeProving/free-compiler into issue-3

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -22,26 +22,75 @@
+--   Both types of η-conversion may also be applied to the same expression.
+--   For example, the function 
+--   
+--  > f :: τ₁ -> τ₂ -> τ₃ > τ 
+--  > f e₁ = case e₁ of {
+--  >           c₁ -> g; 
+--  >           c₂ -> h 
+--  >   }
+--
+--  where @c₁@ and @c₂@ are the constructors of @τ₁@, 
+--  @g@ is a binary function of type @τ₂ -> τ₃ > τ@, and
+--  @h@ is a unary function of type @τ₂ -> (τ₃ > τ)@, 
+--  will be converted to 
```
I just noticed that the dash is missing from some function arrows.

```suggestion
--  > f :: τ₁ -> τ₂ -> τ₃ -> τ
--  > f e₁ = case e₁ of {
--  >           c₁ -> g;
--  >           c₂ -> h 
--  >   }
--
--  where @c₁@ and @c₂@ are the constructors of @τ₁@,
--  @g@ is a binary function of type @τ₂ -> τ₃ -> τ@, and
--  @h@ is a unary function of type @τ₂ -> (τ₃ -> τ)@,
--  will be converted to
```

While writing the suggestion above, I've also noticed that there seem to be some trailing spaces. I'd recommend to configure your editor to remove trailing whitespace automatically.

just95 commented on src/lib/FreeC/Pass/EtaConversionPass.hs:
```diff
@@ -22,26 +22,75 @@
+--  @h@ is a unary function of type @τ₂ -> (τ₃ > τ)@, 
+--  will be converted to 
+-- 
+--  > f e₁ x = case e₁ of
+--  >           c₁ -> \y -> g x y
+--  >           c₂ -> h x 
```
My [previous comment](https://github.com/FreeProving/free-compiler/pull/46#discussion_r440710789) about using IR notation in examples also applies to this code block. It may also be a good idea to write the type annotations in IR notation to demonstrate that `modifyTopLevel` manipulates the type annotations.

```haskell
f (e₁ :: τ₁) :: τ₂ -> τ₃ -> τ = case e₁ of {
    c₁ -> g;
    c₂ -> h 
  }

f (e₁ :: τ₁) (x :: τ₂) :: τ₃ -> τ = case e₁ of {
    c₁ -> \y -> g x y
    c₂ -> h x 
  }
```

Commit:
Improve documentation #3
